### PR TITLE
[TRSL-522] log errors to /opt/ml/output/failure

### PR DESCRIPTION
**Pull request by larroy**
### Description of changes:

When running in SGM, we want to log the errors in  /opt/ml/output/failure so we can have metrics and alarms.

#### Style and formatting:

I have run `pre-commit install && pre-commit run --all-files` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

TRSL-522

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/core/test_errors.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Standard Library
+import os
+import tempfile
+
+# First Party
+from smdebug.exceptions import InvalidCollectionConfiguration
+
+# Local
+from ..util import EnvManager
```
Comment (NihalHarish) [inline tests/core/test_errors.py]: nit: can you replace the relative import with `from tests.util import EnvManager`

Comment (larroy) [reply tests/core/test_errors.py]: sure

Comment (larroy) [reply tests/core/test_errors.py]: tests/core/test_errors.py:None (tests/core/test_errors.py)
ImportError while importing test module '/Users/pllarroy/devel/tornasole/sagemaker-debugger/tests/core/test_errors.py'.
Hint: make sure your test modules/packages have valid Python names.

File: smdebug/core/logger.py [hunk 2]
```diff
@@ -60,6 +60,16 @@ def get_logger(name="smdebug"):
             datefmt="%Y-%m-%d %H:%M:%S",
         )
 
+        if "SM_TRAINING_ENV" in os.environ:
```
Comment (Vikas-kum) [inline smdebug/core/logger.py]: use this function: https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/sagemaker_utils.py#L5

File: smdebug/core/logger.py [hunk 3]
```diff
@@ -60,6 +60,16 @@ def get_logger(name="smdebug"):
...
+            # TRSL-522
+            sm_log_fh = logging.FileHandler(
+                os.environ.get("SMDEBUG_SM_LOGFILE", "/opt/ml/output/failure")
```
Comment (Vikas-kum) [inline smdebug/core/logger.py]: Can you put the details what is expected from SM, once the logs go this file?

File: smdebug/core/logger.py [hunk 4]
```diff
@@ -60,6 +60,16 @@ def get_logger(name="smdebug"):
...
+            )
+            sm_log_fh.setLevel(
+                logging.getLevelName((os.environ.get("SMDEBUG_SM_LOGLEVEL", "INFO")))
```
Comment (Vikas-kum) [inline smdebug/core/logger.py]: will log level Info and higher will go to output/failure?

File: smdebug/exceptions.py [hunk 5]
```diff
@@ -1,58 +1,71 @@
 # First Party
+from smdebug.core.logger import get_logger
 from smdebug.core.modes import ModeKeys as modes
 
 
-class InvalidCollectionConfiguration(Exception):
+class SMDebugError(Exception):
+    def __init__(self):
+        logger = get_logger("exceptions")
```
Comment (Vikas-kum) [inline smdebug/exceptions.py]: logger name is exceptions ?

File: smdebug/exceptions.py [hunk 6]
```diff
@@ -1,58 +1,71 @@
...
+        logger.error("%s: %s", self.__class__.__name__, str(self))
```
Comment (Vikas-kum) [inline smdebug/exceptions.py]: Not all exceptions are "errors" - Example NoMoreData, StepNotAvailable yet.

File: smdebug/exceptions.py [hunk 7]
```diff
@@ -1,58 +1,71 @@
...
+        super().__init__()
 
     def __str__(self):
-        return f"Collection {self.c_name} has not been configured. \
-        Please fill in tensor_name or include_regex"
+        return f"Collection {self.c_name} has not been configured. Please fill in tensor_name or include_regex"
 
 
-class StepNotYetAvailable(Exception):
+class StepNotYetAvailable(SMDebugError):
```
Comment (Vikas-kum) [inline smdebug/exceptions.py]: Depending on what we want to achieve from /opt/../failure, we need to review this, what is treated as error in logs and what is info.

File: tests/util.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# Standard Library
+import os
+
+
+class EnvManager(object):
```
Comment (Vikas-kum) [inline tests/util.py]: Is this file reqd. We have tests which are setting and unsetting env variables. We should reuse the same way.

**Comment by larroy** [comment]
Closing as per feedback from the platform this is no longer needed.

**Pull request closed without merging**